litebird_sim/simulations.py

```diff
@@ -395,26 +395,32 @@ def init_random(self, random_seed):
 
         This function creates a random number generator and saves it in the
         field `random`. It should be used whenever a random number generator
-        is needed in the simulation. It ensures that different MPI processes
-        have their own different seed, which stems from the parameter `random_seed`.
-        The generator is PCG64, and it is ensured that the sequences in
-        each MPI process are independent.
+        is needed in the simulation.
+        In the case `random_seed` has not been set to `None`, it ensures that
+        different MPI processes have their own different seed, which stems
+        from the parameter `random_seed`. The generator is PCG64, and it is
+        ensured that the sequences in each MPI process are independent.
+        In the case `random_seed` has been set to `None`, the generator used
+        is `numpy`'s `default_rng`.
 
         This method is automatically called in the constructor, but it can be
         called again as many times as required. The typical case is when
         one wants to use a seed that has been read from a parameter file.
         """
-        from numpy.random import Generator, PCG64, SeedSequence
+        if random_seed is None:
+            self.random = np.random.default_rng()
+        else:
+            from numpy.random import Generator, PCG64, SeedSequence
 
-        # We need to assign a different random number generator to each MPI
-        # process, otherwise noise will be correlated. The following code
-        # works even if MPI is not used
+            # We need to assign a different random number generator to each MPI
+            # process, otherwise noise will be correlated. The following code
+            # works even if MPI is not used
 
-        # Create a list of N seeds, one per each MPI process
-        seed_seq = SeedSequence(random_seed).spawn(self.mpi_comm.size)
+            # Create a list of N seeds, one per each MPI process
+            seed_seq = SeedSequence(random_seed).spawn(self.mpi_comm.size)
 
-        # Pick the seed for this process
-        self.random = Generator(PCG64(seed_seq[self.mpi_comm.rank]))
+            # Pick the seed for this process
+            self.random = Generator(PCG64(seed_seq[self.mpi_comm.rank]))
```

Comment:
Mmm, if I understand correctly the code, it seems that when the user sets `random_seed=None`, the random generator is no longer PCG. This means that the code is less robust when used with MPI, because it skips the call to `SeedSequence`.

I would rather keep everything as is and only change the call to `SeedSequence` so that if `random_seed=None` it passes a random number to `SeedSequence` (e.g., `np.random.default_rng().integers(0, 2**31)`). In this way, we are sure that the generator will *always* be PCG64.

Reply:
Ah, great, then I guess it's ok to pass `None` to `SeedSequence`. The new code is nice, in this way it's clear we're *always* using PCG64 in any case by just reading the code.

Thanks!

Reply:
[Here](https://docs.scipy.org/doc/numpy-1.3.x/reference/generated/numpy.random.seed.html) I found this:

> If seed is None, then RandomState will try to read data from /dev/urandom (or the Windows analogue) if available or seed from the clock otherwise.

That I guess is something nice? Or could it be dangerous? I mean, if different CPUs look in the same place at the "same time" (maybe it's a naive sentence). I'm not sure what each CPU does... But yeah, I guess it's ok.

Reply:
I'm very sure that this will give different RNGs on different tasks. The intent of `None` here seems to be to produce a truly random RNG, and the random devices on Linux are really good at this.
(Sorry for replying so late, I'm travelling.)

Reply:
ok nice:) thanks:)